### PR TITLE
[Master] Fix stackoverflow when iterating an empty table

### LINKED ISSUE #36406: Iterating over an empty table throws an StackOverflowException
**Description:**
```ballerina
import ballerina/test;

type User record {
    readonly string username;
    string firstName;
};

@test:Config
function tableRemoveProblem() {
    table<User> key(username) users = table [
        {username: "user1", firstName: "John"},
        {username: "user2", firstName: "Jane"}
    ];

    // Remove the members
    _ = users.remove("user1");
    _ = users.remove("user2");

    foreach var user in users {
        test:assertFail("User found " + user.toString());
    }
}
```

Below scenario where we change the removing order; does not crash with an SOE

```ballerina

import ballerina/test;

type User record {
    readonly string username;
    string firstName;
};

@test:Config
function tableRemoveProblem() {
    table<User> key(username) users = table [
        {username: "user1", firstName: "John"},
        {username: "user2", firstName: "Jane"}
    ];

    // Remove the members
    _ = users.remove("user2");
    _ = users.remove("user1");

    foreach var user in users {
        test:assertFail("User found " + user.toString());
    }
}
```

**Steps to reproduce:**

**Affected Versions:**
v2201.0.4 
v2201.1.0

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gimantha**
Fixes #36406

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/langlib-test/src/test/resources/test-src/tablelib_test.bal [hunk 1]
```diff
@@ -683,6 +683,64 @@ function testRemoveThenIterate() returns boolean {
...
+    Employee[] ar = [];
+    var rm1 = data.remove("Mary");
+    var rm2 = data.remove("John");
+    var rm3 = data.remove("Jim");
+
+    Employee newEmp = { name: "JesB", department: "Security" };
+    foreach var v in data {
+        ar.push(v);
+    }
+    data.add(newEmp);
```
Comment (ushirask) [inline langlib/langlib-test/src/test/resources/test-src/tablelib_test.bal:740]: Should we move initializing `newEmp` to here (or inline it) for clarity?

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibTableTest.java [hunk 2]
```diff
@@ -451,6 +451,24 @@ public void testRemoveThenIterate() {
         Assert.assertTrue((Boolean) returns);
     }
 
+    @Test
+    public void testRemoveEmptyThenIterate() {
```
Comment (lasinicl) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibTableTest.java:455]: Shall we use a `dataProvider` instead when invoking the functions?
Also, we can do the assertions in ballerina itself?

File: langlib/langlib-test/src/test/resources/test-src/tablelib_test.bal [hunk 3]
```diff
@@ -683,6 +683,64 @@ function testRemoveThenIterate() returns boolean {
     return ar.length() == 2 && ar[0].name == "John" && ar[1].name == "Jim";
 }
 
+function testRemoveEmptyThenIterate() returns boolean {
```
Comment (pcnfernando) [inline langlib/langlib-test/src/test/resources/test-src/tablelib_test.bal:686]: Shall we add a test iterating an empty table with query?

**Review by pcnfernando — APPROVED** [review]

**Review by lasinicl — APPROVED** [review]

**Review by ushirask — DISMISSED** [review]

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java [hunk 4]
```diff
@@ -496,7 +496,7 @@ public Object next() {
 
         @Override
         public boolean hasNext() {
-           return cursor < noOfAddedEntries;
+           return cursor < noOfAddedEntries && values.size() != 0;
         }
```
Comment (chiranSachintha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java:500]:
```ballerina
type User record {
    readonly string username;
    string firstName;
};

// @test:Config
function tableRemoveProblem() {
    table<User> key(username) users = table [
        {username: "user1", firstName: "John"},
        {username: "user2", firstName: "Jane"}
    ];

    // Remove the members
    _ = users.remove("user1");
    _ = users.remove("user2");
    foreach var user in users {
    }
}

public function main() {
    tableRemoveProblem();
}
```
IMO this is not the correct fix. when considering the above example,  after removing `user1` and `user2`, `noOfAddedEntries` should be equal to zero.
In here https://github.com/ballerina-platform/ballerina-lang/blob/1538dc17bc415ebe947df159f17cbad79b80830f/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java#L675
we do a check before  `noOfAddedEntries` decrement. Do we need this check?
@gimantha WDYT?

Comment (gimantha) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java:500]: There are some requirements from iterator implementation coming from the spec. You cannot remove the members which are not already iterated. Any mutations to members since the iterator is created should panic. To track the point where the current cursor is at, the `noOfAddedEntries` is used. Thats not an exact representation of the existing number of members in the table.

**Review by chiranSachintha — APPROVED** [review]

File: language-server/modules/langserver-core/src/test/resources/completion/module_var_context/config/config17.json [hunk 5]
```diff
@@ -128,6 +128,39 @@
       "sortText": "B",
       "insertText": "TestClass1",
       "insertTextFormat": "Snippet"
+    },
```
Comment (gimantha) [inline language-server/modules/langserver-core/src/test/resources/completion/module_var_context/config/config17.json:131]: @IMS94 This change was done to fix a build failure in master

File: language-server/modules/langserver-core/src/test/resources/semantictokens/config/project/main.bal [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import ballerina/iox;
 import ballerina/httpx;
-import ballerina/log;
+import ballerina/logx;
```
Comment (gimantha) [inline language-server/modules/langserver-core/src/test/resources/semantictokens/config/project/main.bal:3]: @IMS94 I had a chat with @malinthar  regarding this. Seems if the log module is ~/.ballerina folder, this import causes issues. So log module was renamed same as the other stdlibs module have been.

**Review by chiranSachintha — APPROVED** [review]
